File: imu_calib_matlab/src/tasks/run_noise_stats.m
function task = run_noise_stats(inputData)
%RUN_NOISE_STATS Run static noise statistics using the minimum required input.
% Minimum input:
%   static.gyro
%   static.acc

staticData = normalize_static_input(inputData);
missing = require_inputs(staticData, {'gyro', 'acc'});
if ~isempty(missing)
    task = make_task_result(false, 'static.gyro and static.acc are required for noise statistics.', [], ...
        'missing_inputs', {'static.gyro', 'static.acc'}, ...
        'meta', struct('task_name', 'run_noise_stats'));
    return;
end

noiseStats = estimate_noise_stats(staticData.gyro, staticData.acc);
task = make_task_result(true, 'noise statistics estimated successfully.', ...
    struct('noiseStats', noiseStats), ...
    'meta', struct('task_name', 'run_noise_stats'));
end

function staticData = normalize_static_input(inputData)
if isstruct(inputData) && isfield(inputData, 'static')
    staticData = inputData.static;
else
    staticData = inputData;
end
end

File: imu_calib_python/imu_calib/tasks/run_noise_stats.py
from __future__ import annotations

from dataclasses import is_dataclass

from imu_calib.calib.estimate_noise_stats import estimate_noise_stats
from imu_calib.tasks.make_task_result import make_task_result
from imu_calib.tasks.require_inputs import require_inputs


def run_noise_stats(input_data) -> dict:
    static = _normalize_static_input(input_data)
    missing = require_inputs(static, ["gyro", "acc"])
    if missing:
        return make_task_result(
            False,
            "static.gyro and static.acc are required for noise statistics.",
            None,
            missing_inputs=["static.gyro", "static.acc"],
            meta={"task_name": "run_noise_stats"},
        )

    noise_stats = estimate_noise_stats(static.gyro, static.acc)
    return make_task_result(
        True,
        "noise statistics estimated successfully.",
        {"noiseStats": noise_stats},
        meta={"task_name": "run_noise_stats"},
    )


def _normalize_static_input(input_data):
    if is_dataclass(input_data) and hasattr(input_data, "static"):
        return input_data.static
    if isinstance(input_data, dict) and "static" in input_data:
        return input_data["static"]
    return input_data
